File: tests/workflow/producer/rust/src/producer/responses/groupa_structa.rs
use super::{identification, producer::Control, protocol, scope::Scope, Context};
use crate::stat::Alias;
use crate::test::samples;
use clibri::server;
use uuid::Uuid;

type BroadcastStructD = (Vec<Uuid>, protocol::StructD);

pub enum Response {
    RootA((protocol::StructA, BroadcastStructD)),
    RootB(protocol::StructB),
}

#[allow(unused_variables)]
pub async fn response<'c, E: server::Error, C: server::Control<E>>(
    request: &protocol::GroupA::StructA,
    scope: &mut Scope<'_, E, C>,
) -> Result<Response, protocol::GroupA::StructB> {
    let index = scope
        .context
        .get_index(scope.identification.uuid(), Alias::GroupAStructA);
    if index == 1 {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::StructA);
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::StructD);
        Ok(Response::RootA((
            samples::struct_a::get(),
            (vec![scope.identification.uuid()], samples::struct_d::get()),
        )))
    } else if index == 2 {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::StructB);
        Ok(Response::RootB(samples::struct_b::get()))
    } else {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::GroupAStructB);
        Err(samples::group_a::struct_b::get())
    }
}

File: tests/workflow/producer/typescript/src/producer/responses/groupa.structa.ts
import { Response } from "../implementation/responses/groupa.structa";
import {
	Context,
	Producer,
	Identification,
	Filter,
	Protocol,
} from "../implementation/responses";
import { Scope } from "../implementation/scope";
import { Alias } from "../../stat";
import { panic } from "../../tools";

export function response(
	request: Protocol.GroupA.StructA,
	scope: Scope
): Promise<Response> {
	const stat = scope.context.getStat(scope.consumer.uuid());
	stat.case(Alias.GroupAStructA);
	const index = stat.index(Alias.GroupAStructA);
	if (index === 0) {
		return Promise.resolve(
			new Response(Protocol.StructA.defaults())
				.broadcast([scope.consumer.uuid()])
				.StructD(Protocol.StructD.defaults())
		);
	} else if (index === 1) {
		return Promise.resolve(new Response(Protocol.StructB.defaults()));
	} else if (index === 2) {
		return Promise.resolve(
			new Response(Protocol.GroupA.StructB.defaults())
		);
	} else {
		panic(`Unexpected index for Protocol.GroupA.StructA request: ${index}`);
		return Promise.reject(
			new Error(`Handler for Protocol.GroupA.StructA isn't implemented.`)
		);
	}
}
